In [1]:
# 정수

a = 1000
print(a)

a = -7
print(a)

a = 0
print(a)

1000
-7
0


In [2]:
# 실수

# 양의 실수
a = 157.12
print(a)


# 음의 실수
a = -157.12
print(a)

# 정수부가 0일 때 0을 생략
a = -.7
print(a) # -0.7

157.12
-157.12
-0.7


In [3]:
# 10억의 지수 표현 방식

a = 1e9
print(a)

# 752.5
a = 75.25e1
print(a)

# 3.954
a = 3954e-3
print(a)

1000000000.0
752.5
3.954


In [6]:
# 실수 표현에 대한 한계

a = 0.3 + 0.6
print(a)

if a == 0.9:
  print(True)
else :
  print(False)

# round 함수 사용해보기
# round(실수형, 소수점 이하 몇 자리까지 남길지(반올림 위치))
print(round(123.456, 2))

a = 0.3 + 0.6
print(round(a, 4))

if round(a, 4) == 0.9:
  print(True)
else :
  print(False)

0.8999999999999999
False
123.46
0.9
True


`print(round(a, 4))`
- a = 0.8999999999999999
- round(a, 4) → 소수점 넷째 자리까지 반올림 → "0.9" 출력됨

In [9]:
a = 7
b = 3

# 자료형 연산

print(a+b)
print(a-b)
print(a*b)

# / 나눗셈은 실수형으로 나온다
print(a/b)

# 나머지
print(a%b)

# 몫
print(a//b)

# 거듭제곱
print(a**2)

10
4
21
2.3333333333333335
1
2
49


In [13]:
# 리스트 자료형

a = [1,2,3,4,5,6,7,8,9]
print(a)

# 인덱싱 (4번째 인덱스 즉, 다섯 번째 원소에 접근)
print(a[4]) # 5

# 빈 리스트 선언 방법_1
a_list = list()
print(a_list)

# 빈 리스트 선언 방법_2
a_list2 = []
print(a_list2)

[1, 2, 3, 4, 5, 6, 7, 8, 9]
5
[]
[]


In [14]:
# 크기가 n이고, 모든 값이 0인 1차원 리스트 초기화

n = 10
a = [0] * n
print(a)

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


`a = [0] * n`
- a = [0] : [0]은 원소 하나(0)만 있는 리스트
- `* n`은 리스트 반복 연산자
- 즉, 길이가 10이고 모든 원소가 0인 리스트를 생성.

In [18]:
# 리스트 인덱싱

a = [1,2,3,4,5,6,7,8,9]
print(a[-1])

print(a[-3])

a[3] = 7
print(a)

# 슬라이싱
print(a[1:4]) # 2,3,7
print(a[::2]) # 1,3,5,7,9

9
7
[1, 2, 3, 7, 5, 6, 7, 8, 9]
[2, 3, 7]
[1, 3, 5, 7, 9]


In [19]:
# 리스트 컴프리헨션

# 0부터 19까지의 수 중에서 홀수만 포함하는 리스트

array = [i for i in range(20) if i % 2 == 1]

print(array)

'''
같은 코드

array = []

for i in range(20):
  if i % 2 == 1:
    array.append(i)

print(array)
'''

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]


`array = [i for i in range(20) if i % 2 == 1]`
- `range(20)` : 0부터 19까지 정수 생성 → [0, 1, 2, ..., 19]

- `for i in range(20)` : 0부터 19까지 하나씩 i에 대입하며 반복.

- `if i % 2 == 1` : i % 2는 i를 2로 나눈 나머지.== 1 → 홀수만 선택.

- `[i for i in range(20) if i % 2 == 1]` :

  - 리스트 내포(list comprehension) 문법.

  - range(20) 안의 홀수만 뽑아 리스트로 만듦.

  ---

## 컴프리헨션(comprehension)
- 파이썬에서 리스트, 집합, 딕셔너리를 간단하고 직관적으로 만드는 문법

### 📌 리스트 컴프리헨션 기본 구조
```python
[표현식 for 변수 in 반복가능객체 if 조건식]
```
- 표현식 → 리스트에 넣고 싶은 값
- for 변수 in 반복가능객체 → 반복문
- if 조건식 → (선택사항) 조건이 맞을 때만 추가

#### 예시
- 0에서 4까지 제곱수를 리스트에 넣고 싶을 때 ...

**원래 반복문**
```python
squares = []
for i in range(5):
    squares.append(i * i)

print(squares)  # [0, 1, 4, 9, 16]
```

**컴프리헨션**
```python
squares = [i * i for i in range(5)]
print(squares)  # [0, 1, 4, 9, 16]
```

In [20]:
# 1부터 9까지의 수의 제곱 값을 포함하는 리스트

array = [i * i for i in range(1,10)]

print(array)

[1, 4, 9, 16, 25, 36, 49, 64, 81]


In [22]:
# N X M 크기의 2차원 리스트 초기화

n = 3
m = 3

array = [[0] * m for _ in range(n)]
print(array)

[[0, 0, 0], [0, 0, 0], [0, 0, 0]]


`[[0] * m for _ in range(n)]`

- `[0] * m → [0, 0, 0]` :  (길이 m인 리스트)

- `for _ in range(n)` → 위 리스트를 n번 반복해서 새로 생성

- 즉, 3×3 크기의 2차원 리스트가 만들어짐.

In [25]:
# 잘못된 초기화 방법

n = 3
m = 3
array = [[0] * m for _ in range(n)]
print(array)

array[1][1] = 5
print(array)


# 내부적으로 포함된 4개의 리스트 모두가 동일한 객체에 대한 3개의 레퍼런스로 인식하기 때문에
# 특정한 크기를 가지는 2차원 리스트를 초기화할 때는 컴프리헨션을 이용하면 좋다

[[0, 0, 0], [0, 0, 0], [0, 0, 0]]
[[0, 0, 0], [0, 5, 0], [0, 0, 0]]


In [34]:
# 리스트 관련 메소드

# 기본 리스트
a = [1,4,3]
print(f'기본 리스트 : {a}')

# 리스트에 맨 뒤에 원소 삽입
a.append(2)
print(f'사입 : {a}')

# 오름차순 정렬
a.sort()
print(f'오름차순 정렬 : {a}')

# 내림차순 정렬
a.sort(reverse=True)
print(f'내림차순 정렬 : {a}')

# 리스트 원소 뒤집기
a.reverse()
print(f'원소 뒤집기 : {a}')

# 특정 인덱스에 데이터 추가하기
a.insert(2,3)
print(f'인덱스 2에 3 추가 : {a}') # [1,2,3,3,4]

# 특정 값인 데이터 개수 세기
print(f'값이 3인 데이터 개수 : {a.count(3)} 개')

# 특정 값 데이터 삭제
a.remove(1)
print(f'값이 1인 데이터 삭제 : {a}')

기본 리스트 : [1, 4, 3]
사입 : [1, 4, 3, 2]
오름차순 정렬 : [1, 2, 3, 4]
내림차순 정렬 : [4, 3, 2, 1]
원소 뒤집기 : [1, 2, 3, 4]
인덱스 2에 3 추가 : [1, 2, 3, 3, 4]
값이 3인 데이터 개수 : 2 개
값이 1인 데이터 삭제 : [2, 3, 3, 4]


## 🕓 각 리스트 함수별 시간 복잡도

1. **append()**
- 시간 복잡도: *O(1)* (상수 시간)
- 리스트 맨 끝에 추가하므로 매우 빠릅니다
```python
pythonmy_list.append(x)  # O(1)
```
2. **insert()**
- 시간 복잡도: *O(n)* (선형 시간)
- 중간에 삽입하면 뒤의 모든 원소를 한 칸씩 이동해야 합니다
```python
pythonmy_list.insert(0, x)     # O(n) - 맨 앞 삽입이 최악
my_list.insert(len(my_list), x)  # O(1) - 맨 뒤는 append와 동일
```
3. **remove()**
- 시간 복잡도: *O(n)* (선형 시간)
- 값을 찾기 위해 리스트를 순회하고, 삭제 후 뒤의 원소들을 이동해야 합니다
```python
pythonmy_list.remove(x)  # O(n) - 값 찾기 + 원소 이동
```

In [35]:
a = [1,2,3,4,5,5,5,]
remove_set = {3,5}

# remove_set에 포함되지 않은 값만 저장
result = [i for i in a if i not in remove_set]
print(result)

[1, 2, 4]


In [38]:
# 문자열
data = 'Hello world'
print(data)

data = "Don't you know \"Python\"?"
print(data)

Hello world
Don't you know "Python"?


In [42]:
# 문자열 연산
a = "Hello"
b = 'World'
print(a+b)

a = 'String'
print(a * 3)

a = 'ABCDEF'
print(a[2:4]) # CD

HelloWorld
StringStringString
CD


In [44]:
# 튜플 자료형
# 한 번 선언된 값은 변경할 수 없다
# 소괄호 사용

a = (1,2,3,4)
print(a)

# a[2] = 7
# print(a)

(1, 2, 3, 4)


In [46]:
# 딕셔너리 자료형
# 키-값
# 순서 없음 : 인덱싱, 슬라이싱 못함

data = dict()

data['사과'] = 'apple'
data['바나나'] = 'banana'
data['코코넛'] = 'coconut'

print(data)

if '사과' in data:
  print('\'사과\'를 키로 가지는 데이터가 존재합니다.')

{'사과': 'apple', '바나나': 'banana', '코코넛': 'coconut'}
'사과'를 키로 가지는 데이터가 존재합니다.


In [48]:
# 딕셔너리 자료형 관련 함수

data = dict()
data['사과'] = 'apple'
data['바나나'] = 'banana'
data['코코넛'] = 'coconut'

# 키 데이터만 담은 리스트
key_list = data.keys()
print(key_list)
print()

# 값 데이터만 담은 리스트
value_list = data.values()
print(value_list)
print()

# 각 키에 따른 값을 하니씩 출력
for key in key_list:
  print(data[key])

dict_keys(['사과', '바나나', '코코넛'])

dict_values(['apple', 'banana', 'coconut'])

apple
banana
coconut


In [50]:
# 집합 자료형
# 중복을 허용하지 않는다
# 순서 없음 : 인덱싱, 슬라이싱 못함
# 값 데이터만 가지게 된다.
# 중괄호 사용

data = set([1,1,2,3,4,4,4])
print(data)

data = {1,1,2,3,4,4,4}
print(data)

{1, 2, 3, 4}
{1, 2, 3, 4}


In [52]:
a = set([1,2,3,4,5])
b = set([3,4,5,6,7])

print(f'합집합 : {a|b}')
print(f'교집합 : {a&b}')
print(f'차집합 : {a-b}')

합집합 : {1, 2, 3, 4, 5, 6, 7}
교집합 : {3, 4, 5}
차집합 : {1, 2}


In [53]:
# 집합 자료형 관련 함수

data = set([1,2,3])
print(data)

# 새로운 원소 추가
data.add(4)
print(data)

# 새로운 원소 여러 개 추가
data.update([5,6])
print(data)

# 특정한 값을 갖는 원소 삭제
data.remove(3)
print(data)

{1, 2, 3}
{1, 2, 3, 4}
{1, 2, 3, 4, 5, 6}
{1, 2, 4, 5, 6}


In [1]:
x = 15


if x >= 10:
  print(x)

15


In [2]:
score = 80

if score >= 90:
  print('A')
elif score >= 80:
  print('B')
elif score >= 70:
  print('C')
else:
  print('F')



B


In [4]:
score = 90

if score >= 70:
  print('성적이 70점 이상입니다.')
  if score >= 90:
    print('성적이 우수합니다')
else :
  print('성적이 70점 미만입니다')
  print('조금 더 분발하세요')

print('프로그램을 종료합니다')

성적이 70점 이상입니다.
성적이 우수합니다
프로그램을 종료합니다


In [5]:
# 비교 연산자
a = 10
b = 20

print(a == b)
print(a == 10)
print(a != b)
print(a > b)
print(a < b)
print(a >= b)
print(a <= b)

False
True
True
False
True
False
True


In [7]:
print(True and True)
print(False and True)
print(True or False)
print(False or False)
print(not(True))
print(not(False))

True
False
True
False
False
True


In [10]:
# in 연산자

list_a = [1,2,3,4]
print(1 in list_a)
print(0 in list_a)

# not in 연산자
print(0 not in list_a)
print(1 not in list_a)

True
False
True
False


In [11]:
socre = 85

if score >= 80:
  pass
else :
  print('성적이 80점 미만입니다.')

print('프로그램을 종료합니다.')

프로그램을 종료합니다.


In [13]:
socre = 85

if score >= 80: result = 'Success'
else: result = 'Fail'

print(result)

Success


In [14]:
score = 85

result = 'Success' if score > 80 else 'Fail'

print(result)

Success


In [15]:
a = [1,2,3,4,5,5,5]
remove_set = {3,4}

result = []
for i in a:
  if i not in remove_set:
    result.append(i)

print(result)

[1, 2, 5, 5, 5]


In [16]:
a = [1,2,3,4,5,5,5]
remove_set = {3,4}

result = [i for i in a if i not in remove_set]
print(result)

[1, 2, 5, 5, 5]


In [18]:
i = 1
result = 0

# i가 9보다 작거나 같을 때 아래 코드를 반복적으로 실행
while i <= 9:
  result += i
  i += 1

print(result)

45


In [20]:
i = 1
result = 0

# i가 9보다 작거나 같을 때 아래 코드를 반복적으로 실행
while i <= 9:
  if i % 2 == 1:
    result += i
  i += 1

print(result)

25


In [21]:
result = 0

# i는 1부터 9까지의 모든 값을 순회
for i in range(1,10):
  result += i

print(result)

45


In [23]:
score = [90,85,77,65,97]

for i in range(5): # 0부터 4까지
  if score[i] >= 80:
    print(i + 1, '번 학생은 합격입니다.')

1 번 학생은 합격입니다.
2 번 학생은 합격입니다.
5 번 학생은 합격입니다.


In [24]:
score = [90,85,77,65,97]
black_list = {2,4}

for i in range(5):
  if i + 1 in black_list:
    continue
  if score[i] >= 80:
    print(i+1, '번 학생은 합격입니다')

1 번 학생은 합격입니다
5 번 학생은 합격입니다


In [25]:
for i in range(2, 10):
  for j in range(1,10):
    print(i, "X", j, "=", i*j)
  print()

2 X 1 = 2
2 X 2 = 4
2 X 3 = 6
2 X 4 = 8
2 X 5 = 10
2 X 6 = 12
2 X 7 = 14
2 X 8 = 16
2 X 9 = 18

3 X 1 = 3
3 X 2 = 6
3 X 3 = 9
3 X 4 = 12
3 X 5 = 15
3 X 6 = 18
3 X 7 = 21
3 X 8 = 24
3 X 9 = 27

4 X 1 = 4
4 X 2 = 8
4 X 3 = 12
4 X 4 = 16
4 X 5 = 20
4 X 6 = 24
4 X 7 = 28
4 X 8 = 32
4 X 9 = 36

5 X 1 = 5
5 X 2 = 10
5 X 3 = 15
5 X 4 = 20
5 X 5 = 25
5 X 6 = 30
5 X 7 = 35
5 X 8 = 40
5 X 9 = 45

6 X 1 = 6
6 X 2 = 12
6 X 3 = 18
6 X 4 = 24
6 X 5 = 30
6 X 6 = 36
6 X 7 = 42
6 X 8 = 48
6 X 9 = 54

7 X 1 = 7
7 X 2 = 14
7 X 3 = 21
7 X 4 = 28
7 X 5 = 35
7 X 6 = 42
7 X 7 = 49
7 X 8 = 56
7 X 9 = 63

8 X 1 = 8
8 X 2 = 16
8 X 3 = 24
8 X 4 = 32
8 X 5 = 40
8 X 6 = 48
8 X 7 = 56
8 X 8 = 64
8 X 9 = 72

9 X 1 = 9
9 X 2 = 18
9 X 3 = 27
9 X 4 = 36
9 X 5 = 45
9 X 6 = 54
9 X 7 = 63
9 X 8 = 72
9 X 9 = 81



In [26]:
# 더하기 함수

def add(a,b):
  return  a+b

print(add(3,7))

10


In [28]:
def add(a,b):
  print(a+b)

add(3,7)

10


In [29]:
def add(a,b):
  print(a+b)

add(b=8, a=4)

12


In [30]:
a = 0

def func():
  global a
  a += 1

for i in range(10):
  func()

print(a)

10


In [32]:

# 람다 표현식 사용
print((lambda a, b:a + b)(3,7))

10


## 람다 표현식 사용
- `lambda a, b: a + b` 는 이름 없는 함수(익명 함수)

- 즉, a와 b를 받아 a + b를 리턴하는 함수 하나를 즉석에서 만든 것

- 뒤에 (3, 7)을 붙이면 → 바로 실행됨

- 결과는 **10**

In [33]:
# 데이터의 개수 입력
n = int(input())

# 각 데이터를 공백으로 구분하여 입력
data = list(map(int, input().split()))

data.sort(reverse=True)
print(data)

3
1 2 3
[3, 2, 1]


In [35]:
# n,m,k 공백으로 구분하여 입력

n,m,k = map(int, input().split())

print(f'n : {n}\nm : {m}\nk : {k}')

1 2 3
n : 1
m : 2
k : 3


In [36]:
import sys

sys.stdin.readline().rstrip()

''

`sys.stdin.readline()`
- 표준 입력(stdin)에서 한 줄을 입력받는 함수.
- input()과 비슷하지만, sys.stdin.readline()이 훨씬 빠름 → 특히 백준 같은 알고리즘 문제에서 많이 씀.
- 입력의 끝에는 보통 \n(줄바꿈)이 같이 들어옴.

`.rstrip()`
- rstrip()은 문자열 오른쪽 끝의 공백 문자(스페이스, 탭, 줄바꿈 등)를 제거.

*정리*
- `sys.stdin.readline().rstrip()` 은
빠른 입력 + 개행 문자 제거 용도로 많이 쓰는 패턴

In [37]:
import sys

# 문자열 입력 받기
data = sys.stdin.readline().rsplit()
print(data)

[]


In [38]:
# 출력할 변수

a = 1
b = 2

print(f'a : {a}\nb : {b}')

a : 1
b : 2


In [39]:
answer = 9

print(f'정답은 {answer} 입니다')

정답은 9 입니다


In [43]:
# 출력할 변수

answer = 7

print('정답은 ' + str(answer) +  ' 입니다')
print('정답은',str(answer),'입니다.')

정답은 7 입니다
정답은 7 입니다.


In [44]:
result = sum([1,2,3,4,5])
print(result)

15


In [45]:
result = min(7,3,5,2)
print(result)

2


In [46]:
result = max(7,3,5,2)
print(result)

7


In [48]:
result = eval("(3+5)*7")
print(result)

56


In [49]:
result = sorted([9,1,8,5,4]) # 오름차순
print(result)

result = sorted([9,1,8,5,4], reverse=True) # 내림차순
print(result)

[1, 4, 5, 8, 9]
[9, 8, 5, 4, 1]


In [50]:
result = sorted([('홍길동', 35), ('이순신',75), ('아무개', 50),], key=lambda x:x[1], reverse=True)
print(result)

[('이순신', 75), ('아무개', 50), ('홍길동', 35)]


```python
result = sorted(
    [('홍길동', 35), ('이순신', 75), ('아무개', 50)],
    key=lambda x: x[1],  # 튜플에서 두 번째 값(점수)으로 정렬
    reverse=True         # 내림차순
)
```
---
1.  sorted 함수
```text
sorted(리스트, key=기준함수, reverse=True/False)
```
- 리스트를 정렬해서 새로운 리스트 반환
- key → 정렬할 때 기준을 정하는 함수
- reverse=True → 내림차순 정렬 (기본은 오름차순)

2. lambda x: x[1]
- 람다(lambda) 표현식 = 이름 없는 함수(익명 함수)
- lambda 매개변수: 반환값
- 예시
```python
f = lambda x: x[1]
print(f(('홍길동', 35)))  # 35
```

In [51]:
data = [9,1,8,5,4]
data.sort()
print(data)

[1, 4, 5, 8, 9]


In [52]:
from itertools import permutations

data = ['A','B','C'] # 데이터 준비

result = list(permutations(data, 3))

print(result)

[('A', 'B', 'C'), ('A', 'C', 'B'), ('B', 'A', 'C'), ('B', 'C', 'A'), ('C', 'A', 'B'), ('C', 'B', 'A')]


In [53]:
from itertools import combinations

data = ['A','B','C'] # 데이터 준비
result = list(combinations(data,2))

print(result)

[('A', 'B'), ('A', 'C'), ('B', 'C')]


In [54]:
from itertools import product

data = ['A','B','C'] # 데이터 준비
result = list(product(data, repeat=2)) # 2개를 뽑는 모든 순열 구하기 (중복 허용)

print(result)

[('A', 'A'), ('A', 'B'), ('A', 'C'), ('B', 'A'), ('B', 'B'), ('B', 'C'), ('C', 'A'), ('C', 'B'), ('C', 'C')]


In [55]:
from itertools import combinations_with_replacement

data = ['A','B','C']
result = list(combinations_with_replacement(data, 2))
print(result)

[('A', 'A'), ('A', 'B'), ('A', 'C'), ('B', 'B'), ('B', 'C'), ('C', 'C')]


In [56]:
import heapq

def heapsort(iterable) :
  h = []
  result = []
  # 모든 원소를 차례대로 힙에 삽입
  for value in iterable :
    heapq.heappush(h, value)
  # 힙에 삽입된 모든 원소를 차례대로 꺼내어 담기
  for i in range(len(h)):
    result.append(heapq.heappop(h))
  return result

result = heapsort([1,3,5,7,9,2,4,6,8,0])
print(result)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


`heapq.heappush(h, value)`  
→ 리스트 h를 **힙으로 취급**하면서, 원소 value를 넣습니다.

- 반복문을 통해 iterable의 모든 값이 최소 힙 h에 저장됩니다.
- 이 과정이 끝나면 h는 힙 구조를 유지하며, 가장 작은 값이 맨 앞(h[0])에 위치합니다.
- 예: [1,3,5,7,...] 넣으면 내부적으로는 정렬과 비슷하게 보이지만 사실은 힙 구조임.

---

## 힙(heap)이란?
- 힙은 **트리(Tree) 구조**의 일종이에요.
그중에서도 **완전 이진 트리(Complete Binary Tree) 형태**를 가집니다.
- **완전 이진 트리**: 위에서부터, 왼쪽부터 차례대로 채워지는 트리.

힙에는 두 가지 규칙이 있어요:
- 1️⃣ 부모 노드 ≤ 자식 노드 (최소 힙, min-heap)
  - “부모가 항상 자식보다 작거나 같다.”
  - 그래서 맨 꼭대기(루트)에는 가장 작은 값이 옵니다.
- 2️⃣ 부모 노드 ≥ 자식 노드 (최대 힙, max-heap)
  - “부모가 항상 자식보다 크거나 같다.”
  - 그래서 맨 꼭대기(루트)에는 가장 큰 값이 옵니다.
- 즉, 최소 힙에서는 **항상 가장 작은 값이 루트(맨 위)**에 존재합니다.

### 그림 예시
예를 들어, 값 [5, 3, 8, 1, 4]가 있다면:

#### 1️⃣ 최소 힙
```markdown
       1
     /   \
    3     8
   / \
  5   4

```

#### 2️⃣ 최대 힙
```markdown
       8
     /   \
    5     3
   / \
  1   4

```

In [2]:
# 파이썬에서는 최대 힙을 제공하지 않음
# 원소의 부호를 임시로 변경하는 방식을 사용해야 함
# 힙에 원소를 삽입하기 전에 잠시 부호를 바꿨다가 힙에서 원소를 꺼낸 뒤에 다시 원소의 부호를 바꾸면 된다.


import heapq

def heapsort(iterable) :
  h = []
  result = []
  # 모든 원소를 차례대로 힙에 삽입
  for value in iterable :
    heapq.heappush(h, -value)
  # 힙에 삽입된 모든 원소를 차례대로 꺼내어 담기
  for i in range(len(h)):
    result.append(-heapq.heappop(h))
  return result

result = heapsort([1,3,5,7,9,2,4,6,8,0])
print(result)

[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


In [3]:
# 이진탐색 bisect
# '정렬된 배열'에서 특정한 원소를 찾아야 할 때 매우 효과적

from bisect import bisect_left, bisect_right

a = [1,2,4,4,8]
x = 4

print(bisect_left(a,x))
print(bisect_right(a,x))

2
4


1. bisect 모듈이란?

- 정렬된 배열에서 원하는 값을 빠르게 찾거나, 끼워 넣을 위치를 알려주는 모듈입니다.
- 내부적으로 이진 탐색(이분 탐색, binary search) 알고리즘을 사용합니다.
- 시간 복잡도는 **O(log n)** → 매우 빠름.

```python
print(bisect_left(a, x))
```
- `bisect_left(a, x)`는 **x가 들어갈 "가장 왼쪽 인덱스"** 를 반환합니다.
- 즉, a에서 x와 같은 값이 여러 개 있다면, 그 중 제일 왼쪽 위치를 알려줍니다.
- 👉 결과: 2  
(리스트 a = [1,2,4,4,8]에서 첫 번째 4는 인덱스 2에 있음)

```python
print(bisect_right(a, x))
```
- `bisect_right(a, x)`는 **x가 들어갈 "가장 오른쪽 인덱스"** 를 반환합니다.
- 즉, a에서 x와 같은 값이 여러 개 있다면, 그 중 제일 오른쪽 위치 + 1을 알려줍니다.
- 👉 결과: 4  
(마지막 4는 인덱스 3에 있고, 그 뒤인 4번째 인덱스 위치를 반환)

In [4]:
from bisect import bisect_left, bisect_right

# 값이 [left_value, right_value]인 데이터의 개수를 반화하는 함수

def country_by_range(a, left_value, right_value) :
  right_index = bisect_right(a, right_value)
  left_index = bisect_left(a, left_value)
  return right_index - left_index

a = [1,2,3,3,3,3,4,4,8]

# 값이 4인 데이터 개수 출력
print(country_by_range(a, 4, 4))

# 값이 [-1,3] 범위에 있는 데이터 개수 출력
print(country_by_range(a, -1, 3))

2
6


## collections
- 자료구조를 제공하는 표준 라이브러리
- 자주 사용되는 클래스 : **deque** , **Counter**
  - **deque** : 큐를 구현
    - 인덱싱, 슬라이싱 등 기능 사용 못함
    - 연속적으로 나열된 데이터의 시작이나 끝부분의 데이터를 삽입하거나 삭제할 때 유용
    - 스택이나 큐의 기능을 모두 포함
    - 첫번째 원소를 제거할 때 : popleft()
    - 마지막 원소를 제거할 때 : pop()
    - 첫번째 인덱스에 원소 x를 삽입할 때 : appendleft(x)
    - 마지막 인덱스에 원소를 삽입할 때 : append(x)

In [5]:
from collections import deque

data = deque([2,3,4])
data.appendleft(1)
data.append(5)

print(data)
print(list(data))

deque([1, 2, 3, 4, 5])
[1, 2, 3, 4, 5]


In [7]:
from collections import Counter
# Counter는 등장 횟수를 세는 기능
# iterable 객체가 주어졌을 때, 해당 객체 내부의 원소가 몇 번씩 등장했는지 알려준다


counter = Counter(['red','blue','red','green','blue','blue'])

print(counter['blue'])
print(counter['green'])
print(dict(counter)) # 딕셔너리 자료형으로 변환

3
1
{'red': 2, 'blue': 3, 'green': 1}


In [9]:
import math

print(math.factorial(5)) # 5 팩토리얼
print(math.sqrt(7)) # 7의 제곱근 출력
print(math.gcd(21,14)) # 최대 공약수
print(math.pi) # 파이 출력
print(math.e) # 자연상수 e 출력

120
2.6457513110645907
7
3.141592653589793
2.718281828459045
